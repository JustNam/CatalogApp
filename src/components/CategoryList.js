import React, { Component } from 'react';
import { Tabs, Tab, TabPanel, TabList } from 'react-web-tabs';
import 'react-web-tabs/dist/react-web-tabs.css';
import ItemList from './ItemList';

class CategoryList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categorySelected: false,
      categoryId: 0,
    };
  }

  componentDidMount() {
    this.props.getCategories();
  }

  getItems(e, categoryId) {
    e.preventDefault();
    this.setState({
      categorySelected: true,
    });
    // eslint-disable-next-line react/destructuring-assignment
    this.props.getItemsInCategory(categoryId);
    this.setState({ categoryId });
  }

  render() {
    // console
    const { category } = this.props;
    const { categoryId, categorySelected } = this.state;
    return (
      <div>
        <Tabs defaultTab="vertical-tab-one" vertical>
          <TabList>
            {category.data.map((category, index) => (
              <Tab
                tabFor={`vertical-tab-${index}`}
                key={category.id}
                onClick={e => this.getItems(e, category.id)}
              >
                <h3>{category.name}</h3>
              </Tab>
            ))}
            <br />
          </TabList>
          {!categorySelected && (
            <div className="welcome">
              Lets's explore our catalog!
              <br />
              Choose a category.
            </div>
          )}
          {category.data.map((category, index) => (
            <TabPanel tabId={`vertical-tab-${index}`} key={category.id}>
              {category.id === categoryId && <ItemList {...this.props} />}
            </TabPanel>
          ))}
        </Tabs>
      </div>
    );
  }
}

export default CategoryList;
