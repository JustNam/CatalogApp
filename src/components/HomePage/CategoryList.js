import React, { Component } from 'react';
import { Tabs, Tab, TabPanel, TabList } from 'react-web-tabs';
import 'react-web-tabs/dist/react-web-tabs.css';
import ItemList from 'components/HomePage/ItemList';
import NavigationBar from 'components/HomePage/NavigationBar';


class CategoryList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categoryId: 0,
    };
  }

  async componentDidMount() {
    // Load the data of default category
    await this.props.getCategories().then(() => {
      const { category } = this.props;
      const { id } = category.data[0];
      this.props.getItemsInCategoryWithPagination(id, 1);
      this.setState({ categoryId: id });
    });
  }

  getItems(e, categoryId) {
    // Get items in chosen category
    e.preventDefault();
    this.props.getItemsInCategoryWithPagination(categoryId, 1);
    this.setState({ categoryId });
  }

  render() {
    // category variable contains the general information about categories
    const { category } = this.props;
    const { categoryId } = this.state;
    return (
      <div>
        <NavigationBar />
        <div className="container page">
          <Tabs defaultTab="vertical-tab-0" vertical>
            <TabList>
              {category.data.map((category, index) => (
                <Tab
                  tabFor={`vertical-tab-${index}`}
                  key={category.id}
                  onClick={e => this.getItems(e, category.id)}
                >
                  <h3>{category.name}</h3>
                </Tab>
              ))}
              <br />
            </TabList>
            {category.data.map((category, index) => (
              <TabPanel tabId={`vertical-tab-${index}`} key={category.id}>
                {category.id === categoryId && <ItemList {...this.props} />}
              </TabPanel>
            ))}
          </Tabs>
        </div>
      </div>
    );
  }
}

export default CategoryList;
