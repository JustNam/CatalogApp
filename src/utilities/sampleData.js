export const errorItem = {
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4)',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const itemWithEmptyCategory = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [],
};

export const noDescriptionItem = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const item = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 3,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 8,
      title: 'Ru3',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply',
      id: 9,
      title: 'Ru2',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const category = {
  data: [
    {
      created_on: '06/26/2019, 08:09:32',
      id: 1,
      name: 'Swimming',
      updated_on: '06/26/2019, 08:09:32',
    },
    {
      created_on: '06/26/2019, 08:09:32',
      id: 2,
      name: 'Running',
      updated_on: '06/26/2019, 08:09:32',
    },
  ],
};
