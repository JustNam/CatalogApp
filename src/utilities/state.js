export const initalCategoryState = {
  data: [],
};

export const initalItemState = {
  data: [],
  categoryId: 0,
  currentPage: 0,
  lastPage: 0,
};

export const initalUserState = {
  loggedIn: localStorage.getItem('accessToken') || false,
};