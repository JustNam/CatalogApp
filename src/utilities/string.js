export const convertCamelCaseToSnakeCase = string => (
  string.replace(/[A-Z]/g, match => `_${match.toLowerCase()}`)
);