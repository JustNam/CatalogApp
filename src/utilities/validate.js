export const validateItemTitle = (title) => {
  const { value } = title;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9\s]+$/)) {
    errors.push('Given title contains invalid character');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};

export const validateUsername = (username) => {
  const { value } = username;
  const errors = [];
  if (!value.match(/^[a-zA-Z0-9]+$/)) {
    errors.push('Given username contains invalid character');
  }
  if (!value.length > 30) {
    errors.push('Given username contains more than 30 characters');
  }
  return {
    value,
    isValid: (errors.length === 0),
    errorMessages: errors,
  };
};