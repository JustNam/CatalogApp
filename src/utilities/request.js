import { convertCamelCaseToSnakeCase } from './string';
import envConfig from '../configurations';

export const convertToRequest = (dict) => {
  const newDict = {};
  Object.keys(dict).forEach((key) => {
    const newKey = convertCamelCaseToSnakeCase(key);
    newDict[newKey] = dict[key];
  });

  return JSON.stringify(newDict);
};

export const callAPI = async (enpoint, method, body, customizedHeaders = {}) => {
  const defaultHeaders = {
    'Content-Type': 'application/json',
  };
  const headers = {
    ...defaultHeaders,
    ...customizedHeaders,
  };

  let tempBody;
  if (body) {
    if (headers['Content-Type'] === 'application/json') {
      // Eliminate fields which is null
      const processedBody = body;
      Object.keys(processedBody).forEach((key) => {
        if (!processedBody[key]) {
          processedBody[key] = '';
        }
      });
      tempBody = convertToRequest(processedBody);
    } else {
      delete headers['Content-Type'];
      tempBody = body;
    }
  } else {
    delete headers['Content-Type'];
  }
  if (localStorage.getItem('accessToken')) {
    headers.authorization = `Bearer ${localStorage.getItem('accessToken')}`;
  }

  const config = {
    method,
    headers,
  };

  // Only allow the other endpoints to assign body
  if (method !== 'GET') {
    config.body = tempBody;
  }

  function handleErrors(response) {
    if (!response.ok) {
      throw response;
    }
    return response;
  }
  const result = await fetch(envConfig.domain.concat(enpoint), config)
    .then(handleErrors)
    .then(response => response.json());
  return result;
};