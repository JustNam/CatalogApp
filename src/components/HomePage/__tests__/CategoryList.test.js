import React from 'react';
import { shallow } from 'enzyme';
import CategoryList from 'components/HomePage/CategoryList';
import { category } from 'utilities/sampleData';

// import { configure } from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';

// configure({ adapter: new Adapter() });
describe('/components/CategoryList', () => {
  let wrapper;
  let props;
  const setup = () => {
    props = {
      category,
      getCategories: jest.fn(() => Promise.resolve()),
      getItems: jest.fn(() => Promise.resolve()),
      getItemsInCategoryWithPagination: jest.fn(() => Promise.resolve()),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<CategoryList {...props} />);
  });

  it('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('It should get the items in first category', () => {
    expect(props.getItemsInCategoryWithPagination).toBeCalled();
  });
  it('It should get the items in another category when another tab is clicked', () => {
    wrapper.find('[tabFor="vertical-tab-0"]').simulate('click', {
      preventDefault: () => {},
    });
    expect(props.getItemsInCategoryWithPagination.mock.calls.length).toBe(2);
  });
});
