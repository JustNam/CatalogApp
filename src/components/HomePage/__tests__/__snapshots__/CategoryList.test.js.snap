// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/CategoryList It should render correctly 1`] = `ShallowWrapper {}`;
